src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php

```diff
@@ -129,6 +132,17 @@ public function process(ContainerBuilder $container)
                         continue;
                     }
 
+                    if ($type && !$p->isOptional() && !class_exists($type) && !interface_exists($type, false)) {
```

Comment:
you should also check whether it allows null, not only whether it is optional. It could allow null without being optional (but then, we can still pass null).

And you should also check the case where `$type` is not a class typehint (scalar type, array, callable, iterable). I'm not sure it can happen here (I don't remember exactly what `ProxyHelper::getTypeHint` does for them.